src/libsyntax/codemap.rs

```diff
@@ -132,13 +130,29 @@ pub struct Span {
     pub expn_id: ExpnId
 }
 
-/// Spans are converted to MultiSpans just before error reporting, either automatically,
-/// generated by line grouping, or manually constructed.
-/// In the latter case care should be taken to ensure that spans are ordered, disjoint,
-/// and point into the same FileMap.
+/// A collection of spans. Spans have two orthogonal attributes:
+///
+/// - they can be *primary spans*. In this case they are the locus of
+///   the error, and would be rendered with `^^^`.
+/// - they can have a *label*. In this case, the label is written next
+///   to the mark in the snippet when we render.
 #[derive(Clone)]
 pub struct MultiSpan {
-    pub spans: Vec<Span>
+    primary_spans: Vec<Span>,
+    span_labels: Vec<(Span, String)>,
```

Comment:
Do you mean "can it just have a field that is a `Vec<SpanLabel>"? I tried that at first but ultimately I found it cleaner this way -- the MultiSpan basically represents the "raw material" from which we will make up a snippet.

At first I tried to have the `push-span` method update a `Vec<SpanLabel>` but the code was pretty messy that way. When you first push, the label is `None` but level is primary, then it becomes `some`, and if another label is added to that same span (not that this seems likely), you would have to go find if there was a previous span with the same value, so as to mark it primary, etc. Not a huge thing, but this felt nicer. I guess if you add same span multiple times, only one might be primary, but that also seems a bit odd -- how do we know which message is the "primary" one? If we were to go that route, I would want to have the `push_label` take a "primary/secondary" boolean argument instead. 

Bah. Sorry if I'm not being clear. TL;DR is that I think there is room to iterate here, we should think about how we want end-users to "build up" the information needed to render a snippet. I found that just `Vec<SpanLabel>` didn't seem right, but I don't know that what is currently there is right either.

(I don't like the name `MultiSpan` though.)